src/pool.rs

```diff
@@ -1217,7 +1230,319 @@ pub fn get_pool(db: &str, user: &str) -> Option<ConnectionPool> {
+            };
+
+            info!("Created a new pool {:?}", pool);
+        }
+    }
+
+    return pool;
+}
+
 /// Get a pointer to all configured pools.
 pub fn get_all_pools() -> HashMap<PoolIdentifier, ConnectionPool> {
     (*(*POOLS.load())).clone()
 }
+
+async fn create_pool_for_proxy(
+    db: &str,
+    psql_user: &str,
+) -> Result<(Option<ConnectionPool>), Error> {
+    let config = get_config();
+    let client_server_map: ClientServerMap = Arc::new(Mutex::new(HashMap::new()));
+
+    let mut new_pools = HashMap::new();
+    let mut address_id: usize = 0;
+
+    let pool_config_opt = config.pools.get_key_value(db);
+
+    let pool_name: &str = pool_config_opt.unwrap().0;
+    let pool_config = pool_config_opt.unwrap().1;
+
+    info!("Created a new pool {:?}", pool_config);
+
+    let mut user: User = pool_config.users.get("0").unwrap().clone();
+    user.username = psql_user.to_string();
+
+    let mut shards = Vec::new();
+    let mut addresses = Vec::new();
+    let mut banlist = Vec::new();
+    let mut shard_ids = pool_config
+        .shards
+        .clone()
+        .into_keys()
+        .collect::<Vec<String>>();
+
+    // Sort by shard number to ensure consistency.
+    shard_ids.sort_by_key(|k| k.parse::<i64>().unwrap());
+    let pool_auth_hash: Arc<RwLock<Option<String>>> = Arc::new(RwLock::new(None));
+
+    for shard_idx in &shard_ids {
+        let shard = &pool_config.shards[shard_idx];
+        let mut pools = Vec::new();
+        let mut servers = Vec::new();
+        let mut replica_number = 0;
+
+        // Load Mirror settings
+        for (address_index, server) in shard.servers.iter().enumerate() {
+            let mut mirror_addresses = vec![];
+            if let Some(mirror_settings_vec) = &shard.mirrors {
+                for (mirror_idx, mirror_settings) in mirror_settings_vec.iter().enumerate() {
+                    if mirror_settings.mirroring_target_index != address_index {
+                        continue;
+                    }
+                    mirror_addresses.push(Address {
+                        id: address_id,
+                        database: db.to_string(),
+                        host: mirror_settings.host.clone(),
+                        port: mirror_settings.port,
+                        role: server.role,
+                        address_index: mirror_idx,
+                        replica_number,
+                        shard: shard_idx.parse::<usize>().unwrap(),
+                        username: user.username.clone(),
+                        pool_name: pool_name.to_string(),
+                        mirrors: vec![],
+                        stats: Arc::new(AddressStats::default()),
+                        error_count: Arc::new(AtomicU64::new(0)),
+                    });
+                    address_id += 1;
+                }
+            }
+
+            let address = Address {
+                id: address_id,
+                database: psql_user.to_string(),
```

Comment:
I think this is an issue : 

It should be something like this : 

database: shard.database.clone(),


Actually when i connect, the username and database are switched : 


2024-12-17T17:59:25.847653Z  WARN ThreadId(06) pgcat::pool: Could not obtain password hashes using auth_query config, ignoring. Error: AuthPassthroughError("Error trying to obtain password from auth_query, ignoring hash for user 'gule_bdv1vtbt6r7g'. Error: ServerStartupError(\"error message\", ServerIdentifier { username: \"postgres\", database: \"gule_bdv1vtbt6r7g\" })")